tensorflow_probability/python/experimental/mcmc/rhat_reducer.py

```diff
@@ -0,0 +1,203 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""RhatReducer for calculating streaming potential scale reduction."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+
+# Dependency imports
+import tensorflow.compat.v2 as tf
+
+from tensorflow_probability.python.experimental.mcmc import reducer as reducer_base
+from tensorflow_probability.python.experimental.stats import sample_stats
+from tensorflow_probability.python.internal import prefer_static as ps
+from tensorflow_probability.python.mcmc.internal import util as mcmc_util
+
+
+__all__ = [
+    'RhatReducer',
+]
+
+
+RhatReducerState = collections.namedtuple(
+    'RhatReducerState', 'sample_shape, rhat_state')
+
+
+class RhatReducer(reducer_base.Reducer):
+  """`Reducer` that computes a running rhat diagnostic statistic.
+
+  `RhatReducer` assumes that all provided chain states include samples from
+  multiple independent Markov chains, and that all of these chains are to
+  be included in the same calculation. `RhatReducer` also assumes that
+  incoming samples have shape `[Ni, Ci1, Ci2,...,CiD]`. Dimension `0` indexes
+  the `Ni > 1` result steps of the Markov Chain. Dimensions `1` through `D`
+  index the `Ci1 x ... x CiD` independent chains to be tested for convergence
+  to the same target. A possible chunking dimension can also be specified via
+  the `axis` parameter of the `one_step` method. This dimension indexes multiple
+  dependent samples from the same Markov chain, but disregards the first sample
+  dimension.
+
+  To illustrate, with no chunking enabled, a chain state of shape `[5, 2, 3, 4]`
+  can be inferred as shape `[2, 3, 4]` samples drawn from 5 independent Markov
+  chains. If chunking is enabled and `axis=1` then the same chain state can be
+  inferred as 3 samples of shape `[2, 4]` from 5 independent Markov chains. Note
+  how the chunking dimension indexes into only the event dimensions, and not the
+  batch dimension.
+
+  As with all reducers, RhatReducer does not hold state information;
+  rather, it stores supplied metadata. Intermediate calculations are held in
+  a state object, which is returned via `initialize` and `one_step` method
+  calls.
+
+  `RhatReducer` is meant to fit into the larger Streaming MCMC framework.
+  `RunningPotentialScaleReduction` in `tfp.experimental.stats` is
+  better suited for more generic streaming covariance needs. More precise
+  algorithmic details can also be found by referencing
+  `RunningPotentialScaleReduction`.
+  """
+
+  def __init__(self, num_parallel_chains, name=None):
+    """Instantiates this object.
+
+    Args:
+      num_parallel_chains: Integer number of independent Markov chains
+        ran in parallel.
```

Comment:
Is there a reason this is different from `tfp.mcmc.potential_scale_reduction_factor`?  There, the integer gives the number of leading Tensor _dimensions_ to be interpreted as independent chains, and the number of chains is inferred from the sizes of those dimensions.

Reply:
Well I was thinking in the context of streaming MCMC, it would always be the leading Tensor dimension that contains the batched dimension/the independent chains (please correct me if I'm wrong) so I didn't see the need of some `independent_chain_ndims` kwarg. In hindsight, I probably don't need the user to provide the `num_parallel_chains` either.